test/test_url_validity.py

```diff
@@ -76,50 +77,77 @@ def detect_lines(diffstr):
     return resultant_lines
 
 
-def check_git_remote_exists(url, version, tags_valid=False):
+def check_git_remote_exists(url, version, tags_valid=False, commits_valid=False):
     """ Check if the remote exists and has the branch version.
     If tags_valid is True query tags as well as branches """
-    cmd = ('git ls-remote %s refs/heads/*' % url).split()
-
-    try:
-        output = subprocess.check_output(cmd).decode('utf-8')
-    except:
-        return False
-    if not version:
-        # If the above passed assume the default exists
-        return True
 
-    if 'refs/heads/%s' % version in output:
-        return True
+    # Check for tags first as they take priority.
+    # From Cloudbees Support:
+    #  >the way git plugin handles this conflict, a tag/sha1 is always preferred to branch as this is the way most user use an existing job to trigger a release build.
+    #  Catching the corner case to #20286
 
-    # If tags are valid. query for all tags and test for version
-    if not tags_valid:
-        return False
+    tag_match = False
     cmd = ('git ls-remote %s refs/tags/*' % url).split()
 
     try:
         output = subprocess.check_output(cmd).decode('utf-8')
-    except:
-        return False
+    except subprocess.CalledProcessError as ex:
+        return (False, 'subprocess call %s failed: %s' % (cmd, ex))
 
     if 'refs/tags/%s' % version in output:
-        return True
-    return False
+        tag_match = True
+    
+    if tag_match:
+        if tags_valid:
+            return (True, '')
+        else:
+            error_str = 'Tags are not valid, but a tag %s was found. ' % version
+            error_str += 'Re: https://github.com/ros/rosdistro/pull/20286'
+            return (False, error_str)
 
+    branch_match = False
+    # check for branch name
+    cmd = ('git ls-remote %s refs/heads/*' % url).split()
 
-def check_source_repo_entry_for_errors(source, tags_valid=False):
+    try:
+        output = subprocess.check_output(cmd).decode('utf-8')
+    except subprocess.CalledProcessError as ex:
+        return (False, 'subprocess call %s failed: %s' % (cmd, ex))
+    if not version:
+        # If the above passed assume the default exists
+        return (True, '')
+
+    if 'refs/heads/%s' % version in output:
+        return (True, '')
+
+    if commits_valid:
+        try:
+            tmpdir = tempfile.mkdtemp()
```

Comment:
This tmpdir needs to get cleaned up.

Reply:
I took a quick look in the tempfile library docs but on the off chance it's there and I didn't find it, is there a context manager based way to mange temp directories a la

```
with tempfile.mkdtemp() as tmpdir
  ....
```

Reply:
e8dc20a